src/Microsoft.FeatureManagement/FeatureFilters/Recurrence/RecurrenceEvaluator.cs

```diff
@@ -0,0 +1,1446 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.FeatureManagement.FeatureFilters
+{
+    static class RecurrenceEvaluator
+    {
+        //
+        // Error Message
+        const string OutOfRange = "The value is out of the accepted range.";
+        const string UnrecognizableValue = "The value is unrecognizable.";
+        const string RequiredParameter = "Value cannot be null.";
+        const string NotMatched = "Start date is not a valid first occurrence.";
+
+        //
+        // Day of week
+        const string Sunday = "Sunday";
+        const string Monday = "Monday";
+        const string Tuesday = "Tuesday";
+        const string Wednesday = "Wednesday";
+        const string Thursday = "Thursday";
+        const string Friday = "Friday";
+        const string Saturday = "Saturday";
+
+        //
+        // Index
+        const string First = "First";
+        const string Second = "Second";
+        const string Third = "Third";
+        const string Fourth = "Fourth";
+        const string Last = "Last";
+
+        //
+        // Recurrence Pattern Type
+        const string Daily = "Daily";
+        const string Weekly = "Weekly";
+        const string AbsoluteMonthly = "AbsoluteMonthly";
+        const string RelativeMonthly = "RelativeMonthly";
+        const string AbsoluteYearly = "AbsoluteYearly";
+        const string RelativeYearly = "RelativeYearly";
+
+        //
+        // Recurrence Range Type
+        const string EndDate = "EndDate";
+        const string Numbered = "Numbered";
+        const string NoEnd = "NoEnd";
+
+        const int WeekDayNumber = 7;
+        const int MinMonthDayNumber = 28;
+        const int MinYearDayNumber = 365;
+
+        /// <summary>
+        /// Checks if a provided timestamp is within any recurring time window specified by the Recurrence section in the time window filter settings.
+        /// If the time window filter has an invalid recurrence setting, an exception will be thrown.
+        /// <param name="time">A time stamp.</param>
+        /// <param name="settings">The settings of time window filter.</param>
+        /// <returns>True if the time stamp is within any recurring time window, false otherwise.</returns>
+        /// </summary>
+        public static bool MatchRecurrence(DateTimeOffset time, TimeWindowFilterSettings settings)
+        {
+            if (settings == null)
+            {
+                throw new ArgumentNullException(nameof(settings));
+            }
+
+            if (!TryValidateSettings(settings, out string paramName, out string reason))
+            {
+                throw new ArgumentException(reason, paramName);
+            }
+
+            if (time < settings.Start.Value)
+            {
+                return false;
+            }
+
```

Comment:
It's kind of confusing why we need to check if `settings.Recurrence` is null. It has already been checked when call `RecurrenceEvaluator.matchRecurrence()` in "TimeWindowFilter.cs"

Reply:
Removed.